### PR TITLE
attention residual support full recompute

### PR DESCRIPTION
**Pull request by xingmingyyj**
### PR Category

User Experience

### PR Types

Others

### Description

attention residual support full recompute

### 是否引起精度变化

是

### CODE REVIEW

File: src/paddlefleet/transformer/block_attn_res.py [hunk 1]
```diff
@@ -66,6 +66,126 @@ class BlockAttnResSublayersSpec:
...
+        all_repr = [*blocks, partial_block]
+
+        with paddle.no_grad():
+            logits_list = []
+            for repr_i in all_repr:
+                variance = (
+                    repr_i.astype("float32").pow(2).mean(axis=-1, keepdim=True)
+                )
+                rms = paddle.sqrt(variance + norm_eps)
+                k_i = repr_i / rms * norm_weight
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/block_attn_res.py:96]: P0 训练态这里固定用 `mean(x^2)` 的 RMSNorm 公式，但该模块的 `norm` 会随 `config.normalization` 构造成 `LayerNorm`；LayerNorm 需要减均值并使用 `bias`，而这些参数在此路径完全被忽略。结果是 `normalization="LayerNorm"` 时训练输出/梯度与 eval 路径及原始 `self.norm(r)` 语义不一致。请限制该 PyLayer 仅用于 RMSNorm 并回退标准路径，或完整复现实际 norm（包括 mean、variance、bias 及 dtype 语义）。

File: tests/single_card_tests/model/test_block_attn_res.py [hunk 2]
```diff
@@ -144,5 +151,191 @@ def test_block_attn_res(self):
...
+        # would still pass with block propagation broken in both paths.
+        self.assertGreater(
+            max(eager_blocks),
+            0,
+            f"no block boundary was crossed: {eager_blocks}",
+        )
+        # The forward pass must consume blocks exactly as it does without
+        # recompute, and the backward re-run must then replay with them too
+        # (extra entries) rather than starting over from an empty list.
+        self.assertEqual(
```
Comment (risemeup1111) [inline tests/single_card_tests/model/test_block_attn_res.py:289]: P1 这个断言与当前 split 实现的重算边界不一致：`_forward_impl_block_attn_res_split_recompute()` 只把 attention 和 MLP 放进 `recompute()`，而 `BlockAttnResFunc.backward()` 直接在 PyLayer 内重算数学，不会再次调用 `BlockAttnRes.forward`。因此 `recompute_blocks` 不会产生这里期待的 `replayed` 条目，测试在正常实现下会因 `assertTrue(replayed)` 失败。请改为验证实际 recompute closure 的执行，或只比较 eager/full-recompute 的 loss、梯度和 block bookkeeping。

File: src/paddlefleet/transformer/transformer_layer.py [hunk 3]
```diff
@@ -699,42 +880,70 @@ def forward(
...
-            )
+            if (
+                self.config.block_attention_residuals
+                and not skip_block_attn_res
+            ):
+                # block_attention_residuals + full_recompute:
+                # attn_res runs outside recompute (PyLayer handles its own
+                # gradient checkpointing); attention and MLP each get their
+                # own recompute wrapper.
+                outputs = self._forward_impl_block_attn_res_split_recompute(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_layer.py:903]: P1 这里切到 split 路径后，前面计算出的 `offload_kwargs` 不再传入任何 `recompute()`；只有未启用 block attention residual 的旧分支仍在使用 `**offload_kwargs`。因此配置 `decoderlayer_act_offload_settings` 命中本层时，开启 `block_attention_residuals + full_recompute` 会静默失去 activation offload，可能直接增加显存峰值并导致原本可运行的训练 OOM。请把 offload 语义透传到对应的 attention/MLP recompute，或明确拒绝该组合并补测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [训练态 PyLayer 归一化](https://github.com/PaddlePaddle/PaddleFleet/pull/1621#discussion_r3699087575) | P0 | ✅ |
| 2 | [full-recompute 重放断言](https://github.com/PaddlePaddle/PaddleFleet/pull/1621#discussion_r3699090314) | P1 | ✅ |
| 3 | [activation offload 丢失](https://github.com/PaddlePaddle/PaddleFleet/pull/1621#discussion_r3699091721) | P1 | ✅ |
| 4 | [PR 描述](https://github.com/PaddlePaddle/PaddleFleet/pull/1621) | P3 | ✅ |
| 5 | [dense FP8 预量化与清理回归](https://github.com/PaddlePaddle/PaddleFleet/pull/1621#discussion_r3699456042) | P0 | ✅ |
| 6 | [mHC selective recompute 被移除](https://github.com/PaddlePaddle/PaddleFleet/pull/1621#discussion_r3699457057) | P1 | ✅ |
| 7 | [Gemma4 origin_input_ids 丢失](https://github.com/PaddlePaddle/PaddleFleet/pull/1621#discussion_r3699458118) | P0 | ✅ |
| 8 | [LayerNorm 回退测试缺口](https://github.com/PaddlePaddle/PaddleFleet/pull/1621#discussion_r3699532970) | P1 | ✅ |

当前所有审查问题均已修复。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/transformer_layer.py [hunk 4]
```diff
@@ -1237,64 +1466,10 @@ def fp8_quant_weight(self, batch_mode=False, quant_transpose=True):
...
-            if id(m) in seen:
-                continue
-            seen.add(id(m))
-            clear_fn = getattr(m, "clear_fp8_quant_weight", None)
-            if clear_fn is not None:
-                clear_fn()
 
     def use_fp8(self):
         if isinstance(self.mlp, MoELayer):
             return self.mlp.use_fp8()
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_layer.py:1523]: P0 dense FP8 的预量化和清理逻辑在这里被整体删除，但 `GPTModel` 仍会在 `src/paddlefleet/models/gpt/gpt_model.py:808-821` 调用每个 `TransformerLayer.clear_fp8_quant_weight()`；非 MoE 层现在没有该方法，清理回调会直接触发 `AttributeError`。同时，`config.fp8` 的 dense Linear 不再执行预量化，训练/推理会退回未缓存路径。请恢复 dense Linear 的量化与 `clear_fp8_quant_weight` 接口（或同步修改所有上层调用并保留 dense FP8 语义），并补充 dense FP8 回归测试。

File: src/paddlefleet/transformer/transformer_layer.py [hunk 5]
```diff
@@ -1358,48 +1533,6 @@ def __init__(
         # idempotent: already-colored base params are skipped.
         self._mark_shared_no_hook_params()
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_layer.py:1587]: P1 这里删除了 `recompute_mhc_forward` 的配置解析和 `RecomputeWithoutOutput` 路径，因此 `recompute_granularity="selective"` 且 `recompute_modules` 包含 `"mhc_forward"` 时会被静默忽略，显存占用和选择性重算语义发生回退；现有 `tests/single_card_tests/ai_edited_test/transformer/test_ai_hyper_connection_transformer_layer.py` 仍直接检查该属性，也会在构造后报 `AttributeError`。请恢复该配置/执行路径，或显式拒绝该配置并同步迁移测试和文档。

File: src/paddlefleet/transformer/transformer_layer.py [hunk 6]
```diff
@@ -2555,10 +2657,7 @@ def _forward_impl(
 
         if isinstance(self.mlp, MoELayer):
             hidden_states, _ = self.mlp(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_layer.py:2785]: P0 Gemma4 的 MoE 调用不再传递 `origin_input_ids`，但 `TransformerLayer.forward` 仍会在实验版本中从 embedding 侧注入该原始序列，`MoELayer`/router 也会用它计算 aux/z-loss 的有效 token denominator。当前实现会退回局部 `input_ids`，在 MTP、CP 或其他 token 重排场景下改变损失归一化和训练目标。请恢复 `origin_input_ids=origin_input_ids` 透传，并补充实验版本下 Gemma4 MoE 的 loss/denominator 对齐测试。

File: src/paddlefleet/transformer/block_attn_res.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/block_attn_res.py:96]: 当前 head 已验证：`BlockAttnRes` 仅在 `norm` 为 `RMSNorm` 时启用训练态 PyLayer，LayerNorm 等其他归一化会回退标准路径，原问题已修复。

File: tests/single_card_tests/model/test_block_attn_res.py (hunk 2 above)
Comment (risemeup1111) [reply tests/single_card_tests/model/test_block_attn_res.py:289]: 当前 head 已验证：full-recompute 测试已移除不会发生的 `replayed` 断言，改为比较 eager/full-recompute 的 loss、梯度和 block bookkeeping，原问题已修复。

File: src/paddlefleet/transformer/transformer_layer.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_layer.py:903]: 当前 head 已验证：`block_attention_residuals + full_recompute + decoderlayer_act_offload_settings` 现在显式抛出 `ValueError`，并有对应测试覆盖，原问题已修复。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: tests/single_card_tests/model/test_block_attn_res_recompute.py [hunk 7]
```diff
@@ -0,0 +1,924 @@
...
+            output_layer_init_method=functools.partial(
+                paddle.nn.init.xavier_uniform_, gain=1.0
+            ),
+            tie_word_embeddings=True,
+            use_qk_norm=True,
+            block_attention_residuals=True,
+            attn_res_block_size=2,
+        )
+
+    def test_training_uses_pylayer(self):
```
Comment (risemeup1111) [inline tests/single_card_tests/model/test_block_attn_res_recompute.py:296]: P1 当前只保留 RMSNorm 的 `_use_pylayer=True` 断言，LayerNorm/非 RMSNorm 回退的行为测试已被删除。生产代码通过 `isinstance(self.norm, RMSNorm)` 选择 PyLayer，而 `WrappedPaddleNorm` 仍支持 `normalization="LayerNorm"`；若该门控再次退化，现有测试不会发现训练态错误地使用 RMSNorm 公式。请保留一个真实 LayerNorm（或等价非 RMSNorm）的前向/反向测试，并断言走标准路径。

File: src/paddlefleet/transformer/transformer_layer.py (hunk 4 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_layer.py:1523]: 当前 head 已重新核验：`fp8_quant_weight` 已恢复 dense Linear 遍历，`clear_fp8_quant_weight` 接口及清理逻辑恢复，`use_fp8` 也恢复 dense 配置语义；该问题已修复。

File: src/paddlefleet/transformer/transformer_layer.py (hunk 5 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_layer.py:1587]: 当前 head 已重新核验：`HyperConnectionTransformerLayer` 已恢复 `recompute_mhc_forward` 配置解析，并在 attention/MLP 中恢复 `RecomputeWithoutOutput` 重算与注册；现有测试覆盖，问题已修复。

File: src/paddlefleet/transformer/transformer_layer.py (hunk 6 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_layer.py:2785]: 当前 head 已重新核验：Gemma4 `_forward_impl` 已恢复 `origin_input_ids` 参数，并向 `MoELayer` 透传；该问题已修复。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: tests/single_card_tests/model/test_block_attn_res_recompute.py (hunk 7 above)
Comment (risemeup1111) [reply tests/single_card_tests/model/test_block_attn_res_recompute.py:296]: 当前 head 已验证：新增真实 LayerNorm 的标准路径前向/反向测试，并断言不使用 PyLayer，原测试缺口已补齐。

**Review by risemeup1111 — APPROVED** [review]

**Review by LiYuRio — APPROVED** [review]
